File: dot_net/03/ShortestPath/ShortestPath/Graph.cs
using System;
using System.Collections.Generic;

namespace ShortestPath
{
    public class Graph
    {
        private String start;
        private String end;
        private IDictionary<String, Int32> distances;
        private IDictionary<String, String> previous;

        public static object GetDataFromInput(String input)
        {
            String[] lines = input.Split(new string[] { "\r\n" }, StringSplitOptions.None);

            var nodes = lines[0].Split(' ');
            var start = lines[1][0];
            var end = lines[1][2];

            Dictionary<String, Dictionary<String, Int32>> map = new Dictionary<string, Dictionary<string, int>>();
            for (int i = 3; i < lines.Length; i++)
            {
                if (lines[i] == "") continue;
                var fields = lines[i].Split(' ');
                if (!map.ContainsKey(fields[0]))
                    map[fields[0]] = new Dictionary<string, int>();
                map[fields[0]].Add(fields[1], Convert.ToInt32(fields[2]));
            }

            return new 
            {
                nodes = nodes,
                start = start,
                end = end,
                map = map,
            };
        }

        public Graph(IList<String> nodes, IDictionary<String, Dictionary<String, Int32>> map, String start, String end)
        {
            
            this.start = start;
            this.end = end;

            this.distances = new Dictionary<string, int>();
            this.previous = new Dictionary<string, string>();
            foreach (var node in nodes)
            {
                this.distances.Add(node, Int32.MaxValue);
                this.previous.Add(node, null);
            }
            this.distances[this.start] = 0;

            while (nodes.Count > 0)
            {
                var closest = popClosestVertex(ref nodes);

                if (closest == null || closest == end)
                    break;

                foreach (var neighbor in map[closest])
                {
                    if ( distances[neighbor.Key] > distances[closest] + map[closest][neighbor.Key])
                    {
                        distances[neighbor.Key] = distances[closest] + map[closest][neighbor.Key];
                        previous[neighbor.Key] = closest;
                    }
                }
            }
        }

        private String popClosestVertex(ref IList<String> nodes)
        {
            var min = Int32.MaxValue;
            String closest = null;
            foreach (var node in nodes)
            {
                if (distances[node] < min)
                {
                    min = distances[node];
                    closest = node;
                }
            }
            nodes.Remove(closest);
            return closest;
        }

        public IList<String> ShortestPath()
        {
            var next = this.end;
            var path = new List<String>() { next };

            if (this.previous[next] == null)
                return new List<String>();

            while ( next != this.start )
            {
                next = this.previous[next];
                path.Add(next);
            }
            path.Reverse();
            return path;
        }

        public Int32 PathCost()
        {
            return this.distances[this.end];
        }

        public String GetOutput()
        {
            return (this.ShortestPath().Count > 0) ?
                String.Join(" ", this.ShortestPath()) + "\r\n\r\n" + this.PathCost().ToString() :
                String.Format("No path was found from {0} to {1}.", this.start, this.end);
        }
    }
}

File: js/03/Graph.js
/*
 * Graph.js
 *
 * @author Ronald Rey Lovera
 * @contact reyronald@gmail.com
 * @date August 13, 2015
 */

var Graph = (function () {
  // Public functions
  Graph.prototype = {
    shortestPath: shortestPath,
    pathCost: pathCost,
    getOutput: getOutput,
  };

  // Public static functions
  Graph.getGraphDataFromInput = getGraphDataFromInput;

  // Private properties
  var start = end = null;
  var distances = [];
  var previous = [];

  return Graph;

  //////////////////////

	/*
	* Public static functions
	*/

  function getGraphDataFromInput(input) {
    var nodes = input[0].split(" ");
    var start = input[1][0];
    var end = input[1][2];

    var map = {};
    for (var i = 3; i < input.length; i++) {
      var fields = input[i].split(" ");
      if (typeof map[fields[0]] === 'undefined')
        map[fields[0]] = {};
      map[fields[0]][fields[1]] = Number(fields[2]);
    }
    return {
      nodes: nodes,
      start: start,
      end: end,
      map: map,
    }
  }

	/*
	* Public functions
	*/

  function Graph(nodes, map, _start, _end) {
    start = _start;
    end = _end;

    // Initialization
    for (var i in nodes) {
      distances[nodes[i]] = Infinity;
      previous[nodes[i]] = null;
    }
    distances[start] = 0;

    // Dijkstra Algorithm 
    var nodesQueue = nodes.slice(); // Array cloning. See http://jsperf.com/new-array-vs-splice-vs-slice/19
    while (nodesQueue) {
      var closestVertex = popClosestVertex(nodesQueue);

      // If there is no path to the current node, or the current node is the end
      if (closestVertex === null || closestVertex === end)
        break;

      for (var i in map[closestVertex]) {
        if (distances[i] > distances[closestVertex] + map[closestVertex][i]) {
          distances[i] = distances[closestVertex] + map[closestVertex][i];
          previous[i] = closestVertex;
        }
      }
    }
  }

  function shortestPath() {
    var next = end;
    var path = [next];

    if (!previous[next]) return [];

    while (next !== start) {
      next = previous[next];
      path.push(next)
    }
    return path.reverse();
  }

  function pathCost() {
    return distances[end];
  }

  function getOutput() {
    return this.shortestPath().length ?
      this.shortestPath().join(" ") + "\r\n\r\n" + this.pathCost() :
      "No path wass found from " + start + " to " + end + ".";
  }

	/*
	* Private functions
	*/

  function popClosestVertex(q) {
    var min = Infinity;
    var key = index = null;
    for (var i in q) {
      if (distances[q[i]] < min) {
        min = distances[q[i]];
        index = i;
        key = q[i];
      }
    }
    q.splice(index, 1);
    return key;
  }

})();

module.exports = Graph;
